### PR TITLE
Support account locking feature for backup codes

### PR DESCRIPTION
**Pull request by AnuradhaSK**
## Purpose
Support account locking feature for backup codes

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 1]
```diff
@@ -812,15 +807,18 @@ private void processValidUserToken(AuthenticationContext context, AuthenticatedU
...
+     * @param authenticatedUser The authenticatedUser.
+     * @return True if the user entered code matches with a backup code.
+     * @throws AuthenticationFailedException Authentication failure exception when retrieving user claim for OTP list.
      */
-    private void checkWithBackUpCodes(AuthenticationContext context, String userToken,
-                                      AuthenticatedUser authenticatedUser) throws AuthenticationFailedException {
+    private boolean checkWithBackUpCodes(AuthenticationContext context, String userToken,
+                                         AuthenticatedUser authenticatedUser) throws AuthenticationFailedException {
 
+        boolean isMatchedWithBackupCode = false;
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java:821]: Shall we change this to `isMatchingToken`? WDYT?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 2]
```diff
@@ -812,15 +807,18 @@ private void processValidUserToken(AuthenticationContext context, AuthenticatedU
...
      *
-     * @param context           the AuthenticationContext
-     * @param userToken         the userToken
-     * @param authenticatedUser the name of authenticatedUser
-     * @throws AuthenticationFailedException
+     * @param context           The AuthenticationContext.
+     * @param userToken         The userToken.
+     * @param authenticatedUser The authenticatedUser.
+     * @return True if the user entered code matches with a backup code.
+     * @throws AuthenticationFailedException Authentication failure exception when retrieving user claim for OTP list.
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java:816]: Shall we update this as follows? 
`If an error occurred while retrieving user claim for OTP list.`
WDYT?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 3]
```diff
@@ -758,28 +758,23 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
         } else if (isLocalUser && "true".equals(SMSOTPUtils.getBackupCode(context))) {
-            checkWithBackUpCodes(context, userToken, authenticatedUser);
+            succeededAttempt = checkWithBackUpCodes(context, userToken, authenticatedUser);
         } else {
+            context.setProperty(SMSOTPConstants.CODE_MISMATCH, true);
+        }
+        if (!succeededAttempt) {
             handleSmsOtpVerificationFail(context);
-            handleCodeMismatch(context);
+            throw new AuthenticationFailedException("Code mismatch.");
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java:772]: Don't we have to set `context.setProperty(SMSOTPConstants.CODE_MISMATCH, true)` here?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 4]
```diff
@@ -758,28 +758,23 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
-
+        boolean succeededAttempt = false;
         if (userToken.equals(contextToken)) {
             processValidUserToken(context, authenticatedUser);
+            succeededAttempt = true;
         } else if (isLocalUser && "true".equals(SMSOTPUtils.getBackupCode(context))) {
-            checkWithBackUpCodes(context, userToken, authenticatedUser);
+            succeededAttempt = checkWithBackUpCodes(context, userToken, authenticatedUser);
         } else {
+            context.setProperty(SMSOTPConstants.CODE_MISMATCH, true);
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java:768]: Can we add the debug log here?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java (hunk 3 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java:772]: If the user token doesn't match with the code received to the mobile or doesn't match with any backup code `context.setProperty(SMSOTPConstants.CODE_MISMATCH, true)` is set.
Addition to those cases, the `succeededAttempt` is failed for the following reasons,

- Code is expired

- The user hasn't any backup code when the user token is not matched with the code sent to the mobile.

So, We don't need to set `context.setProperty(SMSOTPConstants.CODE_MISMATCH, true)` here. 
I think we can change the error message. eg: `SMS OTP verification failure.` WDYT?

Comment (somindatommy) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java:772]: AFAIU if the code has expired the program will not reach this point right? Therefore this line will only reach for a code mismatch right?
Seems like other methods handle `context.setProperty(SMSOTPConstants.CODE_MISMATCH, true)`. Therefore, we do not need to add it here.
Therefore, shall we change the message something like `Invalid code. Verification failed`? WDYT?

Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java:772]: Yes for SMS OTP, if the code expired it doesn't reach here, but in email OTP it happens.
+1 for `Invalid code. Verification failed`

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java (hunk 4 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java:771]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-sms-otp/pull/94/commits/a8b29a5973113ffb46c61ce362ed50d6877eb259

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java (hunk 3 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: Fixed in a8b29a5

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java (hunk 2 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: Fixed in a8b29a5

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java (hunk 1 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: Fixed in a8b29a5

**Pull request merged**